base/lock.jl

```diff
@@ -164,6 +164,69 @@ function lock(f, l::AbstractLock)
     end
 end
 
+"""
+Lockable(value, lock = ReentrantLock())
+
+Creates a `Lockable` object that wraps `value` and
+associates it with the provided `lock`.
+
+!!! compat "Julia 1.5"
+    Requires at least Julia 1.5.
+"""
+struct Lockable{T, L<:AbstractLock}
+    value::T
+    lock::L
+end
+
+"""
+
+Lockable(value)
+
+Creates a `Lockable` object that wraps `value` and
+associates it with a newly created `ReentrantLock`.
+
+!!! compat "Julia 1.5"
+    Requires at least Julia 1.5.
+"""
+Lockable(value) = Lockable(value, ReentrantLock())
+
+"""
+    lock(f::Function, l::Lockable)
+
+Acquire the lock associated with `l`, execute `f` with the lock held,
+and release the lock when `f` returns. `f` will receive one positional
+argument: the value wrapped by `l`. If the lock is already locked by a
+different task/thread, wait for it to become available.
+
+When this function returns, the `lock` has been released, so the caller should
+not attempt to `unlock` it.
+
+!!! compat "Julia 1.5"
+    Requires at least Julia 1.5.
+"""
+function lock(f, l::Lockable)
+    lock(l.lock) do
+        f(l.value)
+    end
+end
+
+function trylock(f, l::Lockable)
+    if trylock(l.lock)
+        try
+            return f(l.value)
+        finally
+            unlock(l.lock)
+        end
+    end
+    return false
+end
```

Comment:
I'm unsure about this as it's not easy for the caller to decide if `f` is called or not reliably.  I think it might be better to just not define `trylock`.

Having said that, I'd suggest something like

```julia
function trylock(f, l::Lockable)
    if trylock(l.lock)
        try
            return Some(f(l.value))
        finally
            unlock(l.lock)
        end
    end
    return nothing
end
```

if we were to define this.

Maybe @quinnj also has some ideas about the implementation?

Reply:
I’d rather just not define `trylock`.

Reply:
Frankly, no one should be using a `Lockable` except in the `lock(lockable) do object` context. After all, that’s the whole point of bundling the lock and the object together.

Reply:
Alright, I no longer define the method `trylock(f, l::Lockable)`.

@tkf Can this be resolved?

Reply:
I don't see the resolve button. Maybe because I did it as a comment? If you (or anyone) see the resolve button please go ahead and do it.